src/main/scala/mesosphere/marathon/core/election/ElectionService.scala

```diff
@@ -43,58 +54,129 @@ trait ElectionService {
   def leaderHostPort: Option[String]
 
   /**
-    * offerLeadership is called to candidate for leadership. It must be called by candidate only once.
-    *
-    * @param candidate is called back once elected or defeated
-    */
-  def offerLeadership(candidate: ElectionCandidate): Unit
-
-  /**
-    * abdicateLeadership is called to resign from leadership. By the time this method returns,
-    * it can be safely assumed the leadership has been abdicated. This method can be called even
-    * if [[offerLeadership]] wasn't called prior to that, and it will result in Marathon stop and JVM shutdown.
+    * abdicateLeadership is called to resign from leadership.
     */
   def abdicateLeadership(): Unit
 
-  /**
-    * Subscribe to leadership change events.
-    *
-    * The given actorRef will initially get the current state via the appropriate
-    * [[LocalLeadershipEvent]] message and will be informed of changes after that.
-    *
-    * Upon becoming a leader, [[LocalLeadershipEvent.ElectedAsLeader]] is published. Upon leadership loss,
-    * [[LocalLeadershipEvent.Standby]] is sent.
-    */
-  def subscribe(self: ActorRef): Unit
+  def leaderStateEvents: Source[LeadershipState, Cancellable]
+  def leaderTransitionEvents: Source[LeadershipTransition, Cancellable]
+}
 
-  /**
-    * Unsubscribe to any leadership change events for the given [[ActorRef]].
-    */
-  def unsubscribe(self: ActorRef): Unit
+class ElectionServiceImpl(
+    hostPort: String,
+    leaderEventsSource: Source[LeadershipState, Cancellable],
+    crashStrategy: CrashStrategy)(implicit system: ActorSystem) extends ElectionService with StrictLogging {
+
+  import ElectionService._
+  @volatile private[this] var lastState: LeadershipState = LeadershipState.Standby(None)
+  implicit private lazy val materializer = ActorMaterializer()
+
+  override def isLeader: Boolean =
+    lastState == LeadershipState.ElectedAsLeader
+
+  override def localHostPort: String = hostPort
+
+  override def leaderHostPort: Option[String] = lastState match {
+    case LeadershipState.ElectedAsLeader =>
+      Some(hostPort)
+    case LeadershipState.Standby(currentLeader) =>
+      currentLeader
+  }
+
+  override def abdicateLeadership(): Unit = {
+    leaderSubscription.cancel()
+  }
+
+  private def initializeStream() = {
+    val localEventListenerSink = Sink.foreach[LeadershipState] { event =>
+      lastState = event
+    }
+    val crashOnLostLeaderSink = Sink.foreach[LeadershipTransition] {
+      case LeadershipTransition.LostLeadership =>
+        logger.error(s"Lost leadership; crashing")
+        crashStrategy.crash()
+      case _ =>
+        ()
+    }
+
+    val leadershipTransitionsBroadcast = Repeater.sink[LeadershipTransition](32, OverflowStrategy.fail)(system.dispatcher)
```

Comment:
Magic number; we really _shouldn't_ see a burst of leader events in normal operation.